### PR TITLE
URI Pattern Validation and Conflict Resolution design doc

### PR DESCRIPTION
**Pull request by sugmanue**
*Issue #, if available:* #1512 #1029

*Description of changes:*

Design proposal to change the way we validate URI patterns to allow conflicting ones while also specifying how the servers must route ambiguous requests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: designs/uri-conflict-resolution.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+Given two ambiguous URI patterns `A` and `B` with segments
+`[A0, …, An]` and `[B0, …, Bm]` with query string literals
+`[AQ0, …, AQp]` and `[BQ0, …, BQq]` (with both `p` and `q` possibly
+zero, i.e., without query string literals), we use the following
+algorithm to compare them, for each index `x` from `0` to `min(n, m)`
+
+1. If `A[x]` and `B[x]` are both literals then continue (the literal values have to be equal otherwise the patterns would not be ambiguous)
+2. If `A[x]` is a literal and `B[x]` is a label then `A` is more specific than `B`, 
+3. If `A[x]` is a non-greedy label and `B[x]` is a greedy label than `A` is more specific than `B`
```
Comment (kubukoz) [inline designs/uri-conflict-resolution.md:82]:
```suggestion
3. If `A[x]` is a non-greedy label and `B[x]` is a greedy label then `A` is more specific than `B`
```

File: designs/uri-conflict-resolution.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+* `/abc/{foo}/cde`
+* `/abc/{bar}/cde`
+
+## Proposal
+
+### Specificity Routing
+
+The proposal is to  accept any non-equivalent URI path patterns and
+also specifying the expected routing algorithm. The routing algorithm
```
Comment (kubukoz) [inline designs/uri-conflict-resolution.md:62]:
```suggestion
also specify the expected routing algorithm. The routing algorithm
```

**Comment by kubukoz** [comment]
I think this addresses the problem well, thank you! I have no questions so far, just a couple spelling change suggestions.

**Comment by guymguym** [comment]
hi @sugmanue this looks good, but I just wonder - wouldn't it be simple enough to assign an optional numeric priority to each conflicting route, so that api owners could solve any issue with the order of their routes. I would just say that the default priority is 0, and only if there is a collision on the same route and priority, then the api is invalid...

**Comment by kubukoz** [comment]
I feel like doing priority with numbers is like good old `z-index` in CSS - at some point you're going to use the number `999` and regret this decision 😅 

having well-defined rules seems more principled.

**Comment by sugmanue** [comment]
> having well-defined rules seems more principled.

+1 on that, and also opens up the possibility of subtle ways of introducing API breaking modes by changing this value.

**Comment by guymguym** [comment]
I respect that you want to define rules, everyone loves rules, but with that approach I would just be worried that some api's will not be able to use it unless they add an `?x-id=OPNAME` query param to disambiguate between operations. See in S3 for example - https://github.com/awslabs/smithy-rs/issues/1012.

**Comment by guymguym** [comment]
@sugmanue , @kubukoz  , thank you for the replies!

The point I still wanted to make is that using rules that reside "outside" the api definition, in a meta schema, is not going to be more backward compatible because changes to the rules could be implemented at different times for clients vs servers, which will introduce incompatibility. Adding  the routing order as information inside the api definition is actually more stable, and means the api definition is self-contained and described the api behavior as a configuration. So if I keep the api definition in my sources I can track the changes of routing as well. Which I see as a better api design than externalizing the routing as rules.

That said, I understand that a priority could sounds too simple and therefore not sophisticated enough, although I think that such simplicity is usually a good sign, and while z-index is criticized it still provides a pragmatic solution to those non common collision cases.

Two others ways to define such an order:
1. rely on the order of operations in the service.operations list itself.
2. a colliding operation could define an http binding property like routeAfter=OPNAME that would mean that this operation will always be after the other OPNAME and allow it to match first.

Thanks. Guy

**Comment by mtdowling** [comment]
This looks great, Manuel! 

I looked at the S3 Smithy model, and I think all of it can be represented by this proposal _except_ for [CopyObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html) which appears to use headers for routing, and assigning number based priorities to routes wouldn't help here. IMO, header-based routing wouldn't be worth the complexity nor something we'd want to proliferate.

<details>
<summary>Smithy validation errors without x-id</summary>

```
ERROR: com.amazonaws.s3#AbortMultipartUpload (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
  49 |     "com.amazonaws.s3#AbortMultipartUpload": {
     |                                              ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#DeleteObject` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#CopyObject (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
1581 |     "com.amazonaws.s3#CopyObject": {
     |                                    ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#PutObject` (/{Bucket}/{Key+}), `com.amazonaws.s3#UploadPartCopy` (/{Bucket}/{Key+}), `com.amazonaws.s3#UploadPart` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#DeleteObject (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
3168 |     "com.amazonaws.s3#DeleteObject": {
     |                                      ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#AbortMultipartUpload` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#GetObject (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
4995 |     "com.amazonaws.s3#GetObject": {
     |                                   ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#ListParts` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#ListParts (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
8504 |     "com.amazonaws.s3#ListParts": {
     |                                   ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#GetObject` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#PutObject (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
11121 |     "com.amazonaws.s3#PutObject": {
     |                                   ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#CopyObject` (/{Bucket}/{Key+}), `com.amazonaws.s3#UploadPartCopy` (/{Bucket}/{Key+}), `com.amazonaws.s3#UploadPart` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#UploadPart (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
13681 |     "com.amazonaws.s3#UploadPart": {
     |                                    ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#CopyObject` (/{Bucket}/{Key+}), `com.amazonaws.s3#PutObject` (/{Bucket}/{Key+}), `com.amazonaws.s3#UploadPartCopy` (/{Bucket}/{Key+})]

ERROR: com.amazonaws.s3#UploadPartCopy (HttpUriConflict)
     @ /Users/dowling/projects/aws-sdk-js-v3/codegen/sdk-codegen/aws-models/s3.json
     |
13701 |     "com.amazonaws.s3#UploadPartCopy": {
     |                                        ^
     = Operation URI, `/{Bucket}/{Key+}`, conflicts with other operation URIs in the same service: [`com.amazonaws.s3#CopyObject` (/{Bucket}/{Key+}), `com.amazonaws.s3#PutObject` (/{Bucket}/{Key+}), `com.amazonaws.s3#UploadPart` (/{Bucket}/{Key+})]
```

</details>

The thing we'd need to make more explicitly supported is placing dynamic query strings in the URI in addition to input parameters. This would remove the `x-id` thing for everything in the S3 model except CopyObject (TBD if/how there's a different solution for CopyObject, e.g., supporting `x-amz-copy-source` in the query string too). Here's a disambiguation example we could use instead of x-id.

```
AbortMultipartUpload: DELETE /{Bucket}/{Key+}?uploadId
DeleteObject:         DELETE /{Bucket}/{Key+}
```

Smithy validation allows this right now, but it's currently undefined behavior. It would be nice to add statements around if/how this works in the proposal.

**Comment by guymguym** [comment]
@mtdowling you can see the summary of unsupported s3 routing cases in this comment - 
https://github.com/awslabs/smithy-rs/issues/1012#issuecomment-1023243176

for example in the case you mentioned about `?uploadId` we cannot match the route based on a query param with variable input value? Here is the longer description -

> These cases were added because a required query param cannot be specified in the route URI when its value is an actual input field (like ?uploadId=123 or ?partNumber=456), and not a fixed api value (like ?list-type=2). In the smithy spec you cannot specify just a key like ?uploadId in the uri to separate those cases because that's just like ?uploadId= and will fail to route and a value is provided. Maybe the spec can allow it with ?uploadId=* ?

**Comment by mtdowling** [comment]
@guymguym: The way I think it could (and maybe even already) works is that servers would route based on static query string parameters regardless of if they have a value (e.g., `?acl` vs `?acl=true`). So in the `?uploadId` example, they could do the routing based on the presence of the query string parameter, and later, after an operation is matched, deserialize based on the query string binding of a member. We'd only ever want to do this for required input parameters since it would be used for routing. We'd also need to ensure that clients are able to handle this case.

**Comment by zeyad001** [comment]
Any update on this?

**Review by AndrewFossAWS — APPROVED** [review]

**Pull request merged**